File: Zig/lru-cache/lru_cache.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// LRU (Least Recently Used) Cache implementation
/// A fixed-size cache that evicts the least recently used entry when full
pub fn LruCache(comptime K: type, comptime V: type) type {
    return struct {
        const Self = @This();
        
        const Node = struct {
            key: K,
            value: V,
            prev: ?*Node,
            next: ?*Node,
        };
        
        allocator: Allocator,
        capacity: usize,
        size: usize,
        head: ?*Node,
        tail: ?*Node,
        map: std.AutoHashMap(K, *Node),
        
        /// Initialize a new LRU cache with the given capacity
        pub fn init(allocator: Allocator, capacity: usize) !Self {
            return Self{
                .allocator = allocator,
                .capacity = capacity,
                .size = 0,
                .head = null,
                .tail = null,
                .map = std.AutoHashMap(K, *Node).init(allocator),
            };
        }
        
        /// Free all allocated memory
        pub fn deinit(self: *Self) void {
            // Free all nodes
            var current = self.head;
            while (current) |node| {
                const next = node.next;
                self.allocator.destroy(node);
                current = next;
            }
            self.map.deinit();
            self.* = undefined;
        }
        
        /// Get a value from the cache
        /// Returns null if not found, moves the item to the front (most recently used)
        pub fn get(self: *Self, key: K) ?V {
            if (self.map.get(key)) |node| {
                // Move to front
                self.moveToFront(node);
                return node.value;
            }
            return null;
        }
        
        /// Put a value into the cache
        /// If the key exists, updates the value and moves to front
        /// If the cache is full, evicts the least recently used item
        pub fn put(self: *Self, key: K, value: V) !void {
            if (self.map.get(key)) |node| {
                // Update existing node
                node.value = value;
                self.moveToFront(node);
                return;
            }
            
            // Create new node
            const node = try self.allocator.create(Node);
            node.* = Node{
                .key = key,
                .value = value,
                .prev = null,
                .next = null,
            };
            
            // Check if we need to evict
            if (self.size >= self.capacity) {
                self.evict();
            }
            
            // Add to front
            self.addToFront(node);
            try self.map.put(key, node);
            self.size += 1;
        }
        
        /// Check if a key exists in the cache
        pub fn contains(self: *Self, key: K) bool {
            return self.map.contains(key);
        }
        
        /// Remove a key from the cache
        pub fn remove(self: *Self, key: K) bool {
            if (self.map.get(key)) |node| {
                self.removeNode(node);
                _ = self.map.remove(key);
                self.allocator.destroy(node);
                self.size -= 1;
                return true;
            }
            return false;
        }
        
        /// Clear all items from the cache
        pub fn clear(self: *Self) void {
            var current = self.head;
            while (current) |node| {
                const next = node.next;
                self.allocator.destroy(node);
                current = next;
            }
            self.map.clearRetainingCapacity();
            self.head = null;
            self.tail = null;
            self.size = 0;
        }
        
        /// Get current size of the cache
        pub fn getSize(self: *Self) usize {
            return self.size;
        }
        
        /// Get the capacity of the cache
        pub fn getCapacity(self: *Self) usize {
            return self.capacity;
        }
        
        /// Get all keys in order from most to least recently used
        pub fn keys(self: *Self, allocator: Allocator) ![]K {
            var result = std.ArrayList(K).init(allocator);
            var current = self.head;
            while (current) |node| {
                try result.append(node.key);
                current = node.next;
            }
            return result.toOwnedSlice();
        }
        
        // Private helper methods
        
        fn moveToFront(self: *Self, node: *Node) void {
            if (self.head == node) return;
            
            // Remove from current position
            self.removeNode(node);
            
            // Add to front
            self.addToFront(node);
        }
        
        fn addToFront(self: *Self, node: *Node) void {
            node.prev = null;
            node.next = self.head;
            
            if (self.head) |head| {
                head.prev = node;
            }
            self.head = node;
            
            if (self.tail == null) {
                self.tail = node;
            }
        }
        
        fn removeNode(self: *Self, node: *Node) void {
            if (node.prev) |prev| {
                prev.next = node.next;
            } else {
                self.head = node.next;
            }
            
            if (node.next) |next| {
                next.prev = node.prev;
            } else {
                self.tail = node.prev;
            }
        }
        
        fn evict(self: *Self) void {
            if (self.tail) |tail| {
                _ = self.map.remove(tail.key);
                self.removeNode(tail);
                self.allocator.destroy(tail);
                self.size -= 1;
            }
        }
        
        /// Iterator over cache entries from most to least recently used
        pub const Iterator = struct {
            current: ?*Node,
            
            pub fn next(self: *Iterator) ?struct { key: K, value: V } {
                if (self.current) |node| {
                    self.current = node.next;
                    return .{ .key = node.key, .value = node.value };
                }
                return null;
            }
        };
        
        /// Get an iterator over cache entries
        pub fn iterator(self: *Self) Iterator {
            return .{ .current = self.head };
        }
    };
}

// Tests
test "LruCache - basic put and get" {
    const allocator = std.testing.allocator;
    var cache = try LruCache(u32, u32).init(allocator, 3);
    defer cache.deinit();
    
    try cache.put(1, 100);
    try cache.put(2, 200);
    try cache.put(3, 300);
    
    try std.testing.expectEqual(@as(?u32, 100), cache.get(1));
    try std.testing.expectEqual(@as(?u32, 200), cache.get(2));
    try std.testing.expectEqual(@as(?u32, 300), cache.get(3));
    try std.testing.expectEqual(@as(usize, 3), cache.getSize());
}

test "LruCache - eviction" {
    const allocator = std.testing.allocator;
    var cache = try LruCache(u32, u32).init(allocator, 2);
    defer cache.deinit();
    
    try cache.put(1, 100);
    try cache.put(2, 200);
    try cache.put(3, 300); // Should evict 1
    
    try std.testing.expectEqual(@as(?u32, null), cache.get(1));
    try std.testing.expectEqual(@as(?u32, 200), cache.get(2));
    try std.testing.expectEqual(@as(?u32, 300), cache.get(3));
}

test "LruCache - LRU order" {
    const allocator = std.testing.allocator;
    var cache = try LruCache(u32, u32).init(allocator, 3);
    defer cache.deinit();
    
    try cache.put(1, 100);
    try cache.put(2, 200);
    try cache.put(3, 300);
    
    _ = cache.get(1); // Access 1, now it's most recently used
    
    try cache.put(4, 400); // Should evict 2 (least recently used)
    
    try std.testing.expectEqual(@as(?u32, 100), cache.get(1));
    try std.testing.expectEqual(@as(?u32, null), cache.get(2));
    try std.testing.expectEqual(@as(?u32, 300), cache.get(3));
    try std.testing.expectEqual(@as(?u32, 400), cache.get(4));
}

test "LruCache - update existing" {
    const allocator = std.testing.allocator;
    var cache = try LruCache(u32, u32).init(allocator, 2);
    defer cache.deinit();
    
    try cache.put(1, 100);
    try cache.put(1, 200); // Update
    
    try std.testing.expectEqual(@as(?u32, 200), cache.get(1));
    try std.testing.expectEqual(@as(usize, 1), cache.getSize());
}

test "LruCache - remove" {
    const allocator = std.testing.allocator;
    var cache = try LruCache(u32, u32).init(allocator, 3);
    defer cache.deinit();
    
    try cache.put(1, 100);
    try cache.put(2, 200);
    
    try std.testing.expectEqual(true, cache.remove(1));
    try std.testing.expectEqual(@as(?u32, null), cache.get(1));
    try std.testing.expectEqual(@as(usize, 1), cache.getSize());
    try std.testing.expectEqual(false, cache.remove(999)); // Non-existent
}

test "LruCache - clear" {
    const allocator = std.testing.allocator;
    var cache = try LruCache(u32, u32).init(allocator, 3);
    defer cache.deinit();
    
    try cache.put(1, 100);
    try cache.put(2, 200);
    try cache.put(3, 300);
    
    cache.clear();
    
    try std.testing.expectEqual(@as(usize, 0), cache.getSize());
    try std.testing.expectEqual(@as(?u32, null), cache.get(1));
}

test "LruCache - contains" {
    const allocator = std.testing.allocator;
    var cache = try LruCache(u32, u32).init(allocator, 2);
    defer cache.deinit();
    
    try cache.put(1, 100);
    
    try std.testing.expectEqual(true, cache.contains(1));
    try std.testing.expectEqual(false, cache.contains(2));
}

test "LruCache - iterator" {
    const allocator = std.testing.allocator;
    var cache = try LruCache(u32, u32).init(allocator, 3);
    defer cache.deinit();
    
    try cache.put(1, 100);
    try cache.put(2, 200);
    try cache.put(3, 300);
    
    var iter = cache.iterator();
    var count: usize = 0;
    while (iter.next()) |_| {
        count += 1;
    }
    try std.testing.expectEqual(@as(usize, 3), count);
}

File: Swift/lru_cache/LRUCache.swift
/**
 * LRU Cache - A Least Recently Used Cache Implementation
 * 
 * A thread-safe, generic LRU (Least Recently Used) cache that automatically
 * evicts the least recently accessed items when the cache reaches its capacity.
 * 
 * Features:
 * - Generic key-value types (Hashable & Equatable)
 * - O(1) average time complexity for get/set operations
 * - Thread-safe with NSLock
 * - Automatic eviction when capacity is reached
 * - Optional expiration time support
 * 
 * Time Complexity:
 * - get(): O(1) average
 * - set(): O(1) average
 * - contains(): O(1) average
 * 
 * Space Complexity: O(n) where n is the capacity
 * 
 * Author: AllToolkit
 * Date: 2026-05-27
 */

import Foundation

/// A doubly linked list node for LRU tracking
fileprivate final class LRUNode<Key, Value> {
    var key: Key
    var value: Value
    var prev: LRUNode?
    var next: LRUNode?
    var expirationTime: Date?
    
    init(key: Key, value: Value, expirationTime: Date? = nil) {
        self.key = key
        self.value = value
        self.expirationTime = expirationTime
    }
}

/// A thread-safe LRU Cache implementation
public final class LRUCache<Key: Hashable, Value> {
    
    // MARK: - Properties
    
    private let capacity: Int
    private var cache: [Key: LRUNode<Key, Value>] = [:]
    private let head: LRUNode<Key, Value>  // Most recently used
    private let tail: LRUNode<Key, Value>  // Least recently used
    private let lock = NSLock()
    private let defaultTTL: TimeInterval?  // Default time-to-live in seconds
    
    /// Current number of items in the cache
    public var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return cache.count
    }
    
    /// Whether the cache is empty
    public var isEmpty: Bool {
        return count == 0
    }
    
    /// All keys currently in the cache
    public var keys: [Key] {
        lock.lock()
        defer { lock.unlock() }
        return Array(cache.keys)
    }
    
    /// All values currently in the cache (most to least recently used)
    public var values: [Value] {
        lock.lock()
        defer { lock.unlock() }
        var result: [Value] = []
        var node = head.next
        while node !== tail {
            result.append(node!.value)
            node = node!.next
        }
        return result
    }
    
    // MARK: - Initialization
    
    /// Creates an LRU cache with the specified capacity
    /// - Parameters:
    ///   - capacity: Maximum number of items to store (must be > 0)
    ///   - defaultTTL: Default time-to-live for items in seconds (optional)
    public init(capacity: Int, defaultTTL: TimeInterval? = nil) {
        precondition(capacity > 0, "Capacity must be greater than 0")
        self.capacity = capacity
        self.defaultTTL = defaultTTL
        
        // Create dummy head and tail nodes
        head = LRUNode(key: Key()!, value: Value()!)
        tail = LRUNode(key: Key()!, value: Value()!)
        head.next = tail
        tail.prev = head
    }
    
    // MARK: - Public Methods
    
    /// Retrieves a value by key, updating its position to most recently used
    /// - Parameter key: The key to look up
    /// - Returns: The value if found and not expired, nil otherwise
    public func get(_ key: Key) -> Value? {
        lock.lock()
        defer { lock.unlock() }
        
        guard let node = cache[key] else {
            return nil
        }
        
        // Check expiration
        if let expTime = node.expirationTime, expTime < Date() {
            removeNode(node)
            cache.removeValue(forKey: key)
            return nil
        }
        
        // Move to front (most recently used)
        moveToHead(node)
        return node.value
    }
    
    /// Sets a value for a key, potentially evicting the least recently used item
    /// - Parameters:
    ///   - key: The key to set
    ///   - value: The value to store
    ///   - ttl: Time-to-live in seconds (overrides defaultTTL)
    public func set(_ key: Key, value: Value, ttl: TimeInterval? = nil) {
        lock.lock()
        defer { lock.unlock() }
        
        let effectiveTTL = ttl ?? defaultTTL
        let expirationTime = effectiveTTL.map { Date().addingTimeInterval($0) }
        
        if let node = cache[key] {
            // Update existing node
            node.value = value
            node.expirationTime = expirationTime
            moveToHead(node)
        } else {
            // Create new node
            let newNode = LRUNode(key: key, value: value, expirationTime: expirationTime)
            cache[key] = newNode
            addToHead(newNode)
            
            // Evict if over capacity
            if cache.count > capacity {
                if let lru = tail.prev, lru !== head {
                    removeNode(lru)
                    cache.removeValue(forKey: lru.key)
                }
            }
        }
    }
    
    /// Checks if a key exists in the cache without updating its position
    /// - Parameter key: The key to check
    /// - Returns: true if the key exists and is not expired
    public func contains(_ key: Key) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        
        guard let node = cache[key] else {
            return false
        }
        
        // Check expiration
        if let expTime = node.expirationTime, expTime < Date() {
            return false
        }
        
        return true
    }
    
    /// Removes a key from the cache
    /// - Parameter key: The key to remove
    /// - Returns: The removed value, or nil if not found
    @discardableResult
    public func remove(_ key: Key) -> Value? {
        lock.lock()
        defer { lock.unlock() }
        
        guard let node = cache.removeValue(forKey: key) else {
            return nil
        }
        removeNode(node)
        return node.value
    }
    
    /// Clears all items from the cache
    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        
        cache.removeAll()
        head.next = tail
        tail.prev = head
    }
    
    /// Gets a value without updating its LRU position
    /// - Parameter key: The key to look up
    /// - Returns: The value if found and not expired, nil otherwise
    public func peek(_ key: Key) -> Value? {
        lock.lock()
        defer { lock.unlock() }
        
        guard let node = cache[key] else {
            return nil
        }
        
        // Check expiration
        if let expTime = node.expirationTime, expTime < Date() {
            return nil
        }
        
        return node.value
    }
    
    /// Removes all expired items from the cache
    /// - Returns: Number of items removed
    @discardableResult
    public func removeExpired() -> Int {
        lock.lock()
        defer { lock.unlock() }
        
        var removed = 0
        let now = Date()
        
        for (key, node) in cache {
            if let expTime = node.expirationTime, expTime < now {
                removeNode(node)
                cache.removeValue(forKey: key)
                removed += 1
            }
        }
        
        return removed
    }
    
    /// Gets the least recently used key
    /// - Returns: The LRU key, or nil if cache is empty
    public func getLRUKey() -> Key? {
        lock.lock()
        defer { lock.unlock() }
        
        guard tail.prev !== head else {
            return nil
        }
        return tail.prev?.key
    }
    
    /// Gets the most recently used key
    /// - Returns: The MRU key, or nil if cache is empty
    public func getMRUKey() -> Key? {
        lock.lock()
        defer { lock.unlock() }
        
        guard head.next !== tail else {
            return nil
        }
        return head.next?.key
    }
    
    // MARK: - Private Helper Methods
    
    private func addToHead(_ node: LRUNode<Key, Value>) {
        node.prev = head
        node.next = head.next
        head.next?.prev = node
        head.next = node
    }
    
    private func removeNode(_ node: LRUNode<Key, Value>) {
        node.prev?.next = node.next
        node.next?.prev = node.prev
    }
    
    private func moveToHead(_ node: LRUNode<Key, Value>) {
        removeNode(node)
        addToHead(node)
    }
}

// MARK: - Convenience Initializers

extension LRUCache {
    /// Creates an LRU cache with a default capacity of 100
    public convenience init() {
        self.init(capacity: 100)
    }
}

// MARK: - Sequence Conformance

extension LRUCache: Sequence {
    public typealias Element = (key: Key, value: Value)
    
    public func makeIterator() -> AnyIterator<Element> {
        var node = head.next
        return AnyIterator {
            guard node !== self.tail, let current = node else {
                return nil
            }
            node = current.next
            return (key: current.key, value: current.value)
        }
    }
}

// MARK: - CustomStringConvertible

extension LRUCache: CustomStringConvertible {
    public var description: String {
        return "LRUCache<\(Key.self), \(Value.self)>(capacity: \(capacity), count: \(count))"
    }
}

// MARK: - Codable Support (when Key and Value are Codable)

extension LRUCache: Codable where Key: Codable, Value: Codable {
    enum CodingKeys: String, CodingKey {
        case capacity
        case items
        case defaultTTL
    }
    
    public convenience init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let capacity = try container.decode(Int.self, forKey: .capacity)
        let items = try container.decode([(key: Key, value: Value)].self, forKey: .items)
        let defaultTTL = try container.decodeIfPresent(TimeInterval.self, forKey: .defaultTTL)
        
        self.init(capacity: capacity, defaultTTL: defaultTTL)
        
        // Insert items in order (first item becomes LRU)
        for item in items {
            set(item.key, value: item.value)
        }
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(capacity, forKey: .capacity)
        try container.encode(Array(self), forKey: .items)
        try container.encodeIfPresent(defaultTTL, forKey: .defaultTTL)
    }
}

// MARK: - Operator Overloads

extension LRUCache {
    /// Subscript access for get/set operations
    public subscript(key: Key) -> Value? {
        get { return get(key) }
        set {
            if let value = newValue {
                set(key, value: value)
            } else {
                remove(key)
            }
        }
    }
}